File: test-suite/cpp/narrowing_buggy.cpp
// test-suite/cpp/narrowing_buggy.cpp
// Partial null guards: the guard branch logs but does not exit, so the
// fall-through path still dereferences the possibly-null pointer.
// Expect narrowing_cpp findings on every dereference below.

#include <cstdio>
#include <cstddef>

struct Session {
    int user_id;
    char name[32];
};

static void log_null(const char *what) {
    std::printf("null pointer: %s\n", what);
}

// Guard branch only logs; the fall-through dereferences p.
int session_user_id(Session *p) {
    if (!p) {
        log_null("session");
    }
    return p->user_id;
}

// Null-comparison guard falls through; the value is dereferenced anyway.
int deref_value(int *p) {
    if (p == nullptr) {
        log_null("value");
    }
    return *p;
}

// Legacy NULL spelling with the same fall-through defect.
const char *session_name(Session *p) {
    if (p == NULL) {
        log_null("name");
    }
    return p->name;
}

File: test-suite/ruby/narrowing_buggy.rb
# frozen_string_literal: true

# test-suite/ruby/narrowing_buggy.rb — partial nil guards (bead D4).
# Every function below keeps falling through after a nil guard whose branch
# does not exit, so the guarded object can still be nil at the call site.
# Expected: ruby.narrowing.partial_nil_guard findings (buggy >= 1).

require 'logger'

LOGGER = Logger.new($stdout)

# Case 1: `if obj.nil?` branch only logs — fallthrough dereference.
def render_user(user)
  if user.nil?
    LOGGER.warn('missing user, rendering default')
  end
  user.profile.name
end

# Case 2: bare `unless obj` truthiness guard that does not exit.
def charge_card(account)
  unless account
    LOGGER.info('no account supplied')
  end
  account.charge!(19.99)
end

# Case 3: modifier-form partial guard.
def touch_session(session)
  LOGGER.debug('session missing') if session.nil?
  session.touch!
end

# Case 4: instance variable guarded by name.
def load_settings
  if @config.nil?
    LOGGER.warn('config missing, using defaults')
  end
  @config.fetch!(:pool_size)
end
